:book: Add vappconfig.md doc (#175)
vmware-tanzu/vm-operator: docs/tutorials/deploy-vm/vappconfig.md

# Deploy a VM with vAppConfig

This page reviews the bootstrap method vAppConfig.

## Description
The vAppConfig bootstrap method is useful for legacy VM images that rely on bespoke, boot-time processes that leverage vAppConfig properties for customizing a guest.

## Purpose

> As a DevOp, you can assign Golang-based template strings to vAppConfig values to send information into the guest that is not known ahead of time, such as the IPAM data settings used to configure the VM's network stack.

Reviewer: ```suggestion
As a DevOps user, you can assign Golang-based template strings to vAppConfig values to send information into the guest that is not known ahead of time, such as the IPAM data settings used to configure the VM's network stack.
```

## Example

> ```

Reviewer: ```suggestion
```yaml
```
☝🏾 For syntax highlighting

apiVersion: vmoperator.vmware.com/v1alpha1
kind: VirtualMachine
metadata:
  name: legacy-vm
  namespace: test-ns
spec:
  className: best-effort-small
  imageName: haproxy-v0.2.0
  powerState: poweredOn
  storageClass: wcpglobal-storage-profile
  vmMetadata:
    secretName: my-secret
    transport: vAppConfig
---
apiVersion: v1
kind: Secret
metadata:
  name: my-secret
  namespace: test-ns
stringData:
  nameservers: "{{ (index .V1alpha1.Net.Nameservers 0) }}"
  hostname: "{{ .V1alpha1.VM.Name }} "                
  management_ip: "{{ (index (index .V1alpha1.Net.Devices 0).IPAddresses 0) }}"
  management_gateway: "{{ (index .V1alpha1.Net.Devices 0).Gateway4 }}"
```

> 1.  :wave: The Golang based template string is a representation of [vm-operator virtualmachinetempl_types v1alpha1 api](https://github.com/vmware-tanzu/vm-operator/blob/25fb865e615d377192a870583ab32973e9fbd32a/api/v1alpha1/virtualmachinetempl_types.go#L4)

Reviewer: Also for the bullets to appear on separate lines, you will need to add a blank line between the two.

2.  :wave: The fields `nameservers`, `hostname`, `managment_ip` and `management_gateway` are derived from image `haproxy-v0.2.0` OVF properties.

> ```

Reviewer: ```suggestion
```xml
```

# Deploy a VM with vAppConfig

This page reviews the bootstrap method vAppConfig.

## Description
The vAppConfig bootstrap method is useful for l.For egacy VM images that rely on bespoke, boot-time processes that leverage vAppConfig properties for customizing a guest.

## Purpose

> As a DevOp user, you can assign Golang-based template strings to vAppConfig values to send information into the guest that is not known ahead of time, such as the IPAM data settings used to configure the VM's network stack.

Reviewer: ```suggestion
As a DevOps user, you can assign Golang-based template strings to vAppConfig values to send information into the guest that is not known ahead of time, such as the IPAM data settings used to configure the VM's network stack.
```

```
:wave: The Golang based template string is a representation of [vm-operator virtualmachinetempl_types v1alpha1 api](https://github.com/vmware-tanzu/vm-operator/blob/25fb865e615d377192a870583ab32973e9fbd32a/api/v1alpha1/virtualmachinetempl_types.go#L4)

:wave: The fields `nameservers`, `hostname`, `managment_ip` and `management_gateway` are derived from image `haproxy-v0.2.0` OVF properties.
```xml
<Property ovf:key="nameservers" ovf:type="string" ovf:userConfigurable="true" ovf:value="1.1.1.1, 1.0.0.1">
      <Label>2.2. DNS</Label>
      <Description>A comma-separated list of IP addresses for up to three DNS servers</Description>
</Property>
<Property ovf:key="hostname" ovf:type="string" ovf:userConfigurable="true" ovf:value="ubuntuguest">
      <Description>Specifies the hostname for the appliance</Description>
</Property>
<Property ovf:key="management_ip" ovf:type="string" ovf:userConfigurable="true">
      <Label>2.3. Management IP</Label>
      <Description>The static IP address for the appliance on the Management Port Group in CIDR format (.For eg. ip/subnet mask bits). This cannot be DHCP.</Description>
</Property>
```

> ## Supporting Template Queries

Author: What do you think if from Guest customization page we link vAppConfig section to this `vappconfig.md`? same goes to cloudinit, sysprep in the future. So in Guest customization page, we don't go into details on how to deploy but focus on the high level summary?

> ## Supporting Template Queries

Reviewer 2: Yep, that sounds good to me too.

| V1alpha1_IP | `func(IP string) string` | Format a static IP address with default netmask CIDR. If IP is not valid, template string won't be parsed. |
| V1alpha1_IPsFromNIC | `func (index int) []string` | List all IPs with CIDR from the ith NIC. if index out of bound, template string won't be parsed. |
| V1alpha1_SubnetMask | `func(cidr string) (string, error)` | Get subnet mask from a CIDR notation IP address and prefix length. |

### `V1alpha1_FormatIP`
1. Format an IP address with network length. A netmask can be either the length, ex. /24, or the decimal notation, ex. 255.255.255.0. Return IP sans CIDR when input is valid.

2. Format an IP address with CIDR: 
    - if input netmask is different with CIDR, replace and return IP with new CIDR. 
    - if input netmask is empty string, return IP sans CIDR. 
    - if input netmask is not valid, return empty string. 

**Note** when OVF only takes in IP sans CIDR, use `V1alpha1_FormatIP` and pass empty string as input netmask:
```yaml
management_ip: '{{ V1alpha1_FormatIP V1alpha1_FirstIP "" }}' # return first non-loopback IP from first NIC without CIDR. For eg, "192.168.1.10".
```

> ### Examples:

Author: Sounds good. I added a vAppConfig with templating queries and also kept this part.
